sql/hive-thriftserver/src/main/java/org/apache/hive/service/auth/HiveAuthFactory.java

```diff
@@ -362,4 +371,34 @@ public static void verifyProxyAccess(String realUser, String proxyUser, String i
     }
   }
 
+  public static boolean needUgiLogin(UserGroupInformation ugi, String principal, String keytab) {
+    return null == ugi || !ugi.hasKerberosCredentials() || !ugi.getUserName().equals(principal) ||
+      !keytab.equals(getKeytabFromUgi());
```

Comment:
`Object.equals` for keytab check ?
I dont think we invoke this when `keytab` is null currently, but would be good to future proof this.